pkg/charts/defaults.go

```diff
@@ -0,0 +1,265 @@
+	ControllerLogLevel               string
+	ControllerCPULimit               string
+	ControllerCPURequest             string
+	ControllerMemoryLimit            string
+	ControllerMemoryRequest          string
+	ControllerUID                    int64
+	EnableExternalProfiles           bool
+	EnableH2Upgrade                  bool
+	GrafanaCPULimit                  string
+	GrafanaCPURequest                string
+	GrafanaMemoryLimit               string
+	GrafanaMemoryRequest             string
+	ImagePullPolicy                  string
+	IdentityCPULimit                 string
+	IdentityCPURequest               string
+	IdentityMemoryLimit              string
+	IdentityMemoryRequest            string
+	IdentityTrustDomain              string
+	IdentityIssuerClockSkewAllowance time.Duration
+	IdentityIssuerIssuanceLifetime   time.Duration
+	OmitWebhookSideEffects           bool
+	PrometheusCPULimit               string
+	PrometheusCPURequest             string
+	PrometheusImage                  string
+	PrometheusMemoryLimit            string
+	PrometheusMemoryRequest          string
+	ProxyAdminPort                   uint
+	ProxyControlPort                 uint
+	ProxyCPULimit                    string
+	ProxyCPURequest                  string
+	ProxyImageName                   string
+	ProxyInboundPort                 uint
+	ProxyInitImageName               string
+	ProxyInitCPULimit                string
+	ProxyInitCPURequest              string
+	ProxyInitMemoryLimit             string
+	ProxyInitMemoryRequest           string
+	ProxyLogLevel                    string
+	ProxyMemoryLimit                 string
+	ProxyMemoryRequest               string
+	ProxyOutboundPort                uint
+	ProxyUID                         int64
+	WebhookFailurePolicy             string
+}
+
+// ReadDefaults read all the default variables from the values.yaml file.
+// If ha is true, values-ha.yaml will be merged into values.yaml.
+// chartDir is the root directory of the Helm chart where values.yaml is.
+func ReadDefaults(chartDir string, ha bool) (*DefaultValues, error) {
+	valuesFiles := []*chartutil.BufferedFile{
+		{Name: helmDefaultValuesFile},
+	}
+
+	if ha {
+		valuesFiles = append(valuesFiles, &chartutil.BufferedFile{
+			Name: helmDefaultHAValuesFile,
+		})
+	}
+
+	if err := filesReader(chartDir, valuesFiles); err != nil {
+		return nil, err
+	}
+
+	// if ha, merge values.yaml and values-ha.yaml
+	if ha {
+		values := chartutil.Values{}
+		for _, valuesFile := range valuesFiles {
+			v, err := chartutil.ReadValues(valuesFile.Data)
+			if err != nil {
+				return nil, err
+			}
+			values.MergeInto(v)
+		}
+
+		return setDefaults(values, ha)
+	}
+
+	// for non-HA, only interested in values.yaml
+	for _, valuesFile := range valuesFiles {
+		if valuesFile.Name == helmDefaultValuesFile {
+			values, err := chartutil.ReadValues(valuesFile.Data)
+			if err != nil {
+				return nil, err
+			}
+			return setDefaults(values, ha)
+		}
```

Comment:
Those two loops look very similar; how about consolidating with something like this?
```diff
iff --git a/pkg/charts/defaults.go b/pkg/charts/defaults.go
index a713b6fd..9dfb0dd7 100644
--- a/pkg/charts/defaults.go
+++ b/pkg/charts/defaults.go
@@ -1,7 +1,6 @@
 package charts
 
 import (
-       "fmt"
        "time"
 
        "k8s.io/helm/pkg/chartutil"
@@ -79,32 +78,19 @@ func ReadDefaults(chartDir string, ha bool) (*DefaultValues, error) {
                return nil, err
        }
 
-       // if ha, merge values.yaml and values-ha.yaml
-       if ha {
-               values := chartutil.Values{}
-               for _, valuesFile := range valuesFiles {
+       values := chartutil.Values{}
+       for _, valuesFile := range valuesFiles {
+               // if ha, merge values.yaml and values-ha.yaml
+               if ha || valuesFile.Name == helmDefaultValuesFile {
                        v, err := chartutil.ReadValues(valuesFile.Data)
                        if err != nil {
                                return nil, err
                        }
                        values.MergeInto(v)
                }
-
-               return setDefaults(values, ha)
-       }
-
-       // for non-HA, only interested in values.yaml
-       for _, valuesFile := range valuesFiles {
-               if valuesFile.Name == helmDefaultValuesFile {
-                       values, err := chartutil.ReadValues(valuesFile.Data)
-                       if err != nil {
-                               return nil, err
-                       }
-                       return setDefaults(values, ha)
-               }
        }
 
-       return nil, fmt.Errorf("Can't find default Helm values file at %s", chartDir)
+       return setDefaults(values, ha)
 }
```

Reply:
Good :eyes:! I think the `if ha || valuesFile.Name == helmDefaultValuesFile` is probably not needed. We can just merge all `values*.yaml` into the `values` struct.

Reply:
If we always merge, doesn't that imply that the HA values will always override the non-HA ones?

Reply:
Not quite, because `values-ha.yaml` is only added when `ha` is `true`.

```
	if ha {
		valuesFiles = append(valuesFiles, &chartutil.BufferedFile{
			Name: helmDefaultHAValuesFile,
		})
	}
```

Reply:
Ah indeed :+1: